teps/0121-unified-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,152 @@
+- Argo Workflow supports timeout per attempt, but [this stackoverflow question](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout) implies the need of "timeout for all retries".
+
+## Proposal
+We propose to support `retries` in TaskRuns, and solidify that `Timeout` is for each Retry attempt in CustomRun by updating [TEP-0069](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md) and the Developer guidance [docs/runs.md](https://github.com/tektoncd/pipeline/blob/0dd9f9719177054bdf60928ce9b4aab4b85af5bd/docs/runs.md). 
+
+### Notes and Caveats
+One caveat is that **it's hard for PipelineRun Controller to know whether retries is implemented correctly by customers**. If customers want to use a specific CustomRun, they need to check (the documentation or even the code) whether the functionality is supported properly. 
+
+An [alternative](#alternatives) is removing support for retries from TaskRun and CustomRuns, make it a concern of PipelineRun.
+
+Another [alternative](#alternatives) is making retries a specialized field in each CustomRun individually instead of a common field for all CustomRuns.
+
+Another caveat is updating `retriesStatus` can lead to Large CRD issue, which can be alleviated by reducing the information we stored in `retriesStatus`, like removing taskSpec etc. We can design on it (similar to [TEP-0100](https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md)) when it becomes an actual issue.
+
+## Design Details
+
+### API Change
+#### New `retries` field in TaskRun.Spec
+
+```golang
+// TaskRunSpec defines the desired state of TaskRun
+type TaskRunSpec struct {
+  // ...
+  Retries int `json:"retries,omitempty"`
+  // ...
+}
+```
+
+Implements Retries in TaskRun reconciler, use `retriesStatus` to decide whether the retries has been exhausted.
+
+We pass [PipelineTask.Retries](https://github.com/tektoncd/pipeline/blob/481def234d1e8715ea5f0299713592f311340286/pkg/apis/pipeline/v1beta1/pipeline_types.go#L187) to TaskRunSpec.Retries and CustomRunSpec.Retries to retry TaskRuns and CustomRuns belonging to a parent PipelineRun.
+
+## Alternatives
+
+### Support retries for TaskRun and CustomRun in PipelineRun Controller
+
+Remove support for retries in TaskRun and CustomRun, make retries a PipelineRun concern; move logic for retries to PipelineRun reconciler and create new Runs at each attempt (remove the RetriesStatus); Pipeline Task Timeout applies to each retry individually.
+
+In this way, we don't need to worry about whether the retries is implemented correctly in CustomRun.
+
+However, TaskRun and CustomRun **can't retry on their own**, as Runtime Objects that can run on their own, we expect they can also be retried to avoid transient errors. There are use cases implying that we need it:
```

Comment:
Could you please separate runs and taskruns use cases here? 

The reason is that we are implementing taskrun retries because we believe there is a use case for it.
We are keeping retries in runs not because we believe there is a use case for it, but because we believe it is important for all runs to support retries.

The reason I'm asking for this is because I want to make it clear that we don't have a precedent for adding new fields to custom run spec just because some custom runs might need them-- the bar is that *all* custom runs need to support the fields we have in there.

Reply:
Thanks for the comment @lbernick. I truly appreciate that you are so responsible about maintaining the correct course of new feature development.

I'm not sure why would we like to separate runs and taskruns? Why would we believe that all custom runs need to support the "retries" field while we don't necessarily need to support `retries` for all Task Runs.

Creating feature requests is not an effortless action. I think people may not bother to create feature requests for stand-alone TaskRuns potentially because they can actually write scripts to retry them. By supporting retries natively, we can group their retry attempts all together, saving them times on writing scripts etc.

Transient errors are everywhere especially in the Cloud Environment, services can be down for a short peirod of time making the entire TaskRun fails. I truly believe retry is something necessary for every runtime objects to avoid flaky errors.

Reply:
https://learn.microsoft.com/en-us/azure/architecture/best-practices/transient-faults#why-do-transient-faults-occur-in-the-cloud explained how common the transient errors are in the Cloud env. It deepened my understanding about the nature of Cloud Services and how to make things more stable. Hope it to be helpful.

Reply:
Hi Lee @lbernick, I'd like to centralized several other related discussions here:

[Discussion 1](https://github.com/tektoncd/community/pull/816#discussion_r976908878)
> I'm still not completely convinced that there's a use case for taskruns to retry on their own-- nobody has requested this feature AFAIK and I'd imagine that most taskruns people want to retry are part of a pipelinerun. However it does seem reasonable that some people might want to retry the taskruns listed here.

[Discussion 2](https://github.com/tektoncd/community/pull/816#discussion_r977083099)
> With the proposed option, I think we should strongly consider removing retriesStatus from CustomRun status (while still keeping retries in the spec). this will allow customRun controllers to implement retries however they want and prevent the pipelinerun controller from making assumptions about how retries are implemented.

Reply:
Thanks Xinru for your thorough responses here!

My concern is mainly about run controllers. I think implementing retries in TaskRuns is a reasonable choice for the reasons you have pointed out in this thread.

The request for separation here was mainly a request for clarification in the TEP, not a difference in implementation.

I'd like to clarify in the TEP that if we have retries in customrun.spec, we expect all customrun controllers (which we don't control) to implement support for retries (or fail if retries are specified or something like that). This is different than with TaskRuns, because there's only one TaskRun controller which we control, meaning we can easily support retries for all instances of TaskRuns. Therefore, I think it's OK to say that we want to implement retries in the TaskRun controller because some TaskRun instances have a need to retry, but not OK to say that we expect all CustomRun controllers to implement retries because some types of CustomRuns need specialized retry logic. My understanding of the agreement in pipelines WG was that we do actually believe that it's necessary for all CustomRun controllers to support retries, so I wanted to make sure that this is clear in the TEP.

The reason I want to make sure this is clear is because I don't want to introduce new fields in the future to customrun.spec because some custom tasks need them (this seems like how podTemplate ended up in v1alpha1.Run spec).

Happy to clarify further or chat offline if helpful! I appreciate your work on this.

Reply:
Thanks for the detailed reply Lee @lbernick. Yeah that's fair. I removed the use cases here to avoid misleading readers that we want the retries for CustomRun only because some of cutomruns need. Instead, we expect all custom runs to implement it.

Reply:
tahnks! could you point me to the changes, I'm not seeing them?

Reply:
Oh didn't push it XD, updated.

Reply:
thanks! sorry, just one more nitpick: in the alternative solution, taskruns can't retry on their own, but custom runs can retry on their own no matter what we do.

Reply:
Oh right. updated.